python/events/src/hydrotools/events/baseflow/eckhardt.py

```diff
@@ -0,0 +1,288 @@
+
+@dataclass
+class BaseflowData:
+    """DataClass containing a pandas.Series of baseflow values with
+        a DateTimeIndex and associated filter parameter values.
+    
+        Parameters
+        ----------
+        values: pandas.Series
+            A pandas.Series of baseflow values with a DateTimeIndex.
+        recession_constant: float
+            Linear reservoir recession constant, a, from Eckhardt (2005, 2008). Must be
+            between 0.0 and 1.0, typically between 0.85 and 0.95 for daily streamflow.
+            Influenced by basin size, precipitation frequency, and other catchment
+            characteristics.
+        maximum_baseflow_index: float
+            The maximum baseflow index, $BFI_{max}$, from Eckhardt (2005, 2008). Must be
+            between 0.0 and 1.0, typically between 0.25 and 0.8 for daily streamflow.
+            Influenced by aquifer porosity, streamflow frequency, and other catchment
+            characteristics.
+    """
+    values: pd.Series
+    recession_constant: float
+    maximum_baseflow_index: float
+
+def linear_recession_analysis(
+        series: npt.NDArray,
+        window: int = 5,
+        maximum_relative_error: float = 0.02,
+        maximum_iterations: int = 100
+    ) -> float:
+    """Performs recession analysis on series assuming a linear reservoir. Using
+        the method from Eckhardt (2008).
+
+        Parameters
+        ----------
+        series: array-type, required
+            A numpy array of streamflow values.
+        window: int, optional, default 5
+            The minimum number of consecutively decreasing values in series
+            that indicate a period of recession.
+        maximum_relative_error: float, optional, default 0.02
+            The maximum allowable error in the recession constant due to
+            random observational errors. Defaults to 0.02 (2.0 %).
+        maximum_iterations: int, optional, default 100
+            Maximum number of times to remove outliers and recompute.
+
+        Returns
+        -------
+        recession_constant: float
+            Linear reservoir recession constant, a, from Eckhardt (2005, 2008). Must be
+            between 0.0 and 1.0, typically between 0.85 and 0.95 for daily streamflow.
+            Influenced by basin size, precipitation frequency, and other catchment
+            characteristics.
+    """
+    # Find decreases in streamflow
+    # Note first decrease is a[i+1] - a[i]
+    decreases = np.diff(series) < 0.0
+
+    # Inspect overlapping windows to identify recession periods
+    # Periods where all values decreased are "recessions"
+    periods = npl.stride_tricks.sliding_window_view(
+        decreases,
+        window_shape=window
+        )
+    recessions = np.all(periods, axis=1)
+
+    # Determine index of recession period central values
+    # Note: Increase index by one because np.diff
+    #   dropped the original first value
+    indices = np.where(recessions)[0] + window // 2 + 1
```

Comment:
Is there a case for ever wanting to use the end of the recession period instead of the middle?

Reply:
I suppose there could be, though that would be a different method of baseflow separation. Having said that, for a sufficiently long time series I expect the difference to be negligible.